sweep/sweeper.go

```diff
@@ -1,234 +1,627 @@
-	currentBlockHeight uint32) (*wire.MsgTx, error) {
+// NOTE: Extreme care needs to be taken that input isn't changed externally.
+// Because it is an interface and we don't know what is exactly behind it, we
+// cannot make a local copy in sweeper.
+func (s *UtxoSweeper) SweepInput(input Input) (chan Result, error) {
+	if input == nil || input.OutPoint() == nil || input.SignDesc() == nil {
+		return nil, errors.New("nil input received")
+	}
 
-	// Generate the receiving script to which the funds will be swept.
-	pkScript, err := s.cfg.GenSweepScript()
+	log.Debugf("Sweep request received: %v", input.OutPoint())
+
+	sweeperInput := &sweepInputMessage{
+		input:      input,
+		resultChan: make(chan Result, 1),
+	}
+
+	// Deliver input to main event loop.
+	select {
+	case s.newInputs <- sweeperInput:
+	case <-s.quit:
+		return nil, fmt.Errorf("sweeper shutting down")
+	}
+
+	return sweeperInput.resultChan, nil
+}
+
+// collector is the sweeper main loop. It processes new inputs, spend
+// notifications and counts down to publication of the sweep tx.
+func (s *UtxoSweeper) collector() error {
+	defer s.wg.Done()
+
+	bestHash, bestHeight, err := s.cfg.ChainIO.GetBestBlock()
 	if err != nil {
-		return nil, err
+		return err
 	}
 
-	// Using the txn weight estimate, compute the required txn fee.
-	feePerKw, err := s.cfg.Estimator.EstimateFeePerKW(confTarget)
+	blockEpochs, err := s.cfg.Notifier.RegisterBlockEpochNtfn(&chainntnfs.BlockEpoch{
+		Height: bestHeight,
+		Hash:   bestHash,
+	})
 	if err != nil {
-		return nil, err
+		return err
 	}
+	defer blockEpochs.Cancel()
+
+	for {
+		select {
+		case input := <-s.newInputs:
+			outpoint := *input.input.OutPoint()
+			pendInput, pending := s.pendingInputs[outpoint]
+			if pending {
+				log.Debugf("Already pending input %v received",
+					outpoint)
+				// Add additional result channel to signal spend
+				// of this input.
+				pendInput.listeners = append(
+					pendInput.listeners, input.resultChan)
+				continue
+			}
+			// Create a new pendingInput and initialize the
+			// listeners slice with the passed in result channel. If
+			// this input is offered for sweep again, the result
+			// channel will be appended to this slice.
+			pendInput = &pendingInput{
+				listeners:        []chan Result{input.resultChan},
+				input:            input.input,
+				minPublishHeight: bestHeight,
+			}
+			s.pendingInputs[outpoint] = pendInput
+
+			// Start watching for spend of this input, either by us
+			// or the remote party. The possibility of a remote
+			// spend is the reason why just registering for conf. of
+			// our sweep tx isn't enough.
+			cancel, err := s.registerSpendNtfn(
+				outpoint,
+				input.input.SignDesc().Output.PkScript,
+				input.input.HeightHint(),
+			)
+			if err != nil {
+				err = fmt.Errorf(
+					"cannot watch for spend: %v",
+					err)
+
+				s.signalAndRemove(&outpoint,
+					Result{Err: err})
+
+				continue
+			}
+			pendInput.ntfnRegCancel = cancel
```

Comment:
just create the pending input struct here?

Reply:
I need to have the struct already in place and added to the `pendingInputs` map, because failure to register the spend notification is relying on that to signal back the error. (`s.signalAndRemove`)